src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -97,6 +102,36 @@ protected RubyEncoding negotiateSameEncodingUncached(
             return firstEncoding;
         }
 
+        @Specialization(guards = {
+                "firstEncoding != secondEncoding",
+                "firstEncoding == UTF_8",
+                "getCodeRange(second) == CR_7BIT"
+        })
+        protected RubyEncoding negotiateUtf8AndCr7Bit(
+                Rope first, RubyEncoding firstEncoding, Rope second, RubyEncoding secondEncoding) {
+            return Encodings.UTF_8;
+        }
+
+        @Specialization(guards = {
+                "firstEncoding != secondEncoding",
+                "firstEncoding == US_ASCII",
+                "getCodeRange(second) == CR_7BIT"
+        })
+        protected RubyEncoding negotiateUSAsciiAndCr7Bit(
+                Rope first, RubyEncoding firstEncoding, Rope second, RubyEncoding secondEncoding) {
+            return Encodings.US_ASCII;
+        }
+
+        @Specialization(guards = {
+                "firstEncoding != secondEncoding",
+                "firstEncoding == BINARY",
+                "getCodeRange(second) == CR_7BIT"
+        })
+        protected RubyEncoding negotiateAscii8BitAndCr7Bit(
+                Rope first, RubyEncoding firstEncoding, Rope second, RubyEncoding secondEncoding) {
+            return Encodings.BINARY;
+        }
```

Comment:
These 3 specializations could become one with `@Cached` and a guard the cached encoding is either US-ASCII, UTF-8 or BINARY. Not sure what's better.
Single specialization might involve less checks when running the node the first time or when there is not already a matching specialization.

Reply:
Yeah, I can do that. I tried combining into a single specialization and it bloated the compiled code a fair bit. But, I think I forgot to cache the result and was just always returning `firstEncoding`. The hardest part is naming the helper function to check the first encoding is one of the three default runtime encodings.

Reply:
Re naming, maybe `isCommonEncoding` or `isStandardEncoding` or so.
`isUSASCIIOrUTF8OrBINARY` does indeed feel unreadable.

Reply:
Thanks. I like those suggestions. I was trying something like `isDefault`, but thought that might confuse people, especially since the defaults are configurable.